src/main/java/eu/europeana/metis/sandbox/service/metrics/MetricsServiceImpl.java

```diff
@@ -0,0 +1,226 @@
+package eu.europeana.metis.sandbox.service.metrics;
+
+import eu.europeana.metis.sandbox.common.Step;
+import eu.europeana.metis.sandbox.dto.report.ProgressInfoDto;
+import eu.europeana.metis.sandbox.entity.DatasetEntity;
+import eu.europeana.metis.sandbox.entity.metrics.ProgressDatasetEntity;
+import eu.europeana.metis.sandbox.entity.metrics.ProgressStepEntity;
+import eu.europeana.metis.sandbox.repository.DatasetRepository;
+import eu.europeana.metis.sandbox.repository.metrics.ProgressDatasetRepository;
+import eu.europeana.metis.sandbox.repository.metrics.ProgressStepRepository;
+import eu.europeana.metis.sandbox.service.dataset.DatasetReportService;
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.util.HashMap;
+import java.util.List;
+import java.util.LongSummaryStatistics;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.boot.actuate.endpoint.annotation.DeleteOperation;
+import org.springframework.boot.actuate.endpoint.annotation.Endpoint;
+import org.springframework.boot.actuate.endpoint.annotation.ReadOperation;
+import org.springframework.stereotype.Service;
+
+@Service
+@Endpoint(id = "sandbox-metrics")
+public class MetricsServiceImpl implements MetricsService {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MetricsServiceImpl.class);
+  private final Map<String, Lock> metricsLocksMap = new ConcurrentHashMap<>();
+
+  private final DatasetRepository datasetRepository;
+
+  private final DatasetReportService datasetReportService;
+
+  private final ProgressDatasetRepository progressDatasetRepository;
+
+  private final ProgressStepRepository progressStepRepository;
+
+  public MetricsServiceImpl(DatasetRepository datasetRepository,
+      DatasetReportService datasetReportService,
+      ProgressDatasetRepository progressDatasetRepository,
+      ProgressStepRepository progressStepRepository) {
+    this.datasetRepository = datasetRepository;
+    this.datasetReportService = datasetReportService;
+    this.progressDatasetRepository = progressDatasetRepository;
+    this.progressStepRepository = progressStepRepository;
+  }
+
+  @ReadOperation
+  @Override
+  public DatasetMetrics datasetMetrics() {
+    Map<String, Object> metricsMap = new ConcurrentHashMap<>();
+    getProgressDatasetMetrics(metricsMap);
+    getProgressStepMetrics(metricsMap);
+    DatasetMetrics datasetMetrics = new DatasetMetrics();
+    datasetMetrics.setDatasetMetricsMap(metricsMap);
+    return datasetMetrics;
+  }
+
+  @DeleteOperation
```

Comment:
Is it ok to include this method in the cleanup scheduler that already exists? so after X time the metrics old metrics cleaned up. Sort of the idea of LRU.

Reply:
Depends the implementation you'll be using.
There are LRUs that when an item is inserted, it will also check if it goes above the threshold and will automatically remove the lru item, so probably we won't need to do manual cleanups. But do have a good look on the existent implementations.